### PR TITLE
8291600: [vectorapi] vector cast op check is not always needed for vector mask cast

### PR DESCRIPTION
**Pull request by XiaohongGong**
Recently we found the performance of "`FIRST_NONZERO`" for double type is largely worse than the other types on x86 when `UseAVX=2`. The main reason is the "`VectorCastL2X`" op is not supported by the backend when the dst element type is `T_DOUBLE`. This makes the check of `VectorCast` op fail before intrinsifying "`VectorMask.cast()`" which is used in the
"`FIRST_NONZERO`" java implementation (see [1]). However, the compiler will not generate the `VectorCast `op for `VectorMask.cast()` if:

 1) the current platform supports the predicated feature
 2) the element size (in bytes) of the src and dst type is the same

So the check of "`VectorCast`" op is needless for such cases. To fix it, this patch:

 1) limits the specified vector cast op check to vectors
 2) adds the relative mask cast op check for VectorMask.cast()
 3) cleans up the unnecessary codes

Here is the performance of "`FIRST_NONZERO`" benchmark [2] on a x86 machine with `UseAVX=2`:
```
Benchmark                          (size) Mode Cnt Before  After   Units
DoubleMaxVector.FIRST_NONZERO       1024  thrpt 15 49.266 2460.886 ops/ms
DoubleMaxVector.FIRST_NONZEROMasked 1024  thrpt 15 49.554 1892.223 ops/ms
```
[1] https://github.com/openjdk/jdk/blob/master/src/jdk.incubator.vector/share/classes/jdk/incubator/vector/DoubleVector.java#L770
[2] https://github.com/openjdk/panama-vector/blob/vectorIntrinsics/test/micro/org/openjdk/bench/jdk/incubator/vector/operation/DoubleMaxVector.java#L246

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8291600](https://bugs.openjdk.org/browse/JDK-8291600): [vectorapi] vector cast op check is not always needed for vector mask cast

### Reviewers
 * [Jie Fu](https://openjdk.org/census#jiefu) (@DamonFool - **Reviewer**)
 * [Eric Liu](https://openjdk.org/census#eliu) (@theRealELiu - Author) ⚠️ Review applies to [cd751795](https://git.openjdk.org/jdk/pull/9737/files/cd751795615a55c669a54245125042df89e6f9cd)
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - **Reviewer**)

### Contributors
 * Quan Anh Mai `<qamai@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9737/head:pull/9737` \
`$ git checkout pull/9737`

Update a local copy of the PR: \
`$ git checkout pull/9737` \
`$ git pull https://git.openjdk.org/jdk pull/9737/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9737`

View PR using the GUI difftool: \
`$ git pr show -t 9737`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9737.diff">https://git.openjdk.org/jdk/pull/9737.diff</a>

</details>

### CODE REVIEW

**Comment by merykitty** [comment]
May I ask if we can use the `VectorMaskCast` nodes for the non-predicated mask casts. As the value of the elements can only be 0 or -1, we can generate better code on avx, as we don't need to truncate the elements first as in the general cases. Thanks a lot.

**Comment by XiaohongGong** [comment]
Thanks a lot for looking at this PR!

> May I ask if we can use the `VectorMaskCast` nodes for the non-predicated mask casts. As the value of the elements can only be 0 or -1, we can generate better code on avx, as we don't need to truncate the elements first as in the general cases. Thanks a lot.

If the element size of src and dst type is equal, we can use `VectorMaskCast` which doesn't need to emit any instructions. But for others like `S -> I`, `S->B`, I'm afraid that we need the extending or narrowing instructions for non-predicated mask casts. 

For example, for a short vector mask A with 128 bits: 
````
A:  0000 1111 0000 1111 0000 1111 0000 1111
````
If we want to cast it to an int vector mask B with 256 bits, the result should be extended to:
```
B: 00000000 11111111 00000000 11111111 00000000 11111111 00000000 11111111
```
And if we want to cast it to a byte vector mask C with 64 bits, the result should be narrowed to:
```
C: 00 11 00 11 00 11 00 11
```

And yes, for such cases, we can also generate a `VectorMaskCast` node and together with the relative backend match rules for it when the element size is different, which I think the codes will be the same with `VectorCast`.  This makes the codes in mid-end cleaner, but may need some duplicate match rules.  I'm not quite familiar with avx instructions, so for such cases, does avx has better instructions than the VectorCast for it? If so, I'd like to always use `VectorMaskCast` for all platforms and add the backend rules if needed. Also considering I'm not familiar with the x86 instructions, the change may needs your help. WDYT?

**Comment by merykitty** [comment]
Yes you are right, the code would be mostly the same, which means we can reuse the existing match rules to additionally match `VectorMaskCast` for those cases. For other cases, in particular, narrowing cast to subword types, since avx < 3 does not support truncation cast and only provides saturation cast instructions. We need to truncate the upper bits ourselves. For example, a cast from `int` to `byte` is done as follow

    vpand dst, src, [external address mask]
    vpackusdw dst, dst
    vpermq dst, dst, 0x08
    vpackuswb dst, dst

For vector mask cast, we can get rid of the first masking and use the `vpackss`s instead, which removes the need to reference memory. Thanks.

**Comment by XiaohongGong** [comment]
> Yes you are right, the code would be mostly the same, which means we can reuse the existing match rules to additionally match `VectorMaskCast` for those cases. For other cases, in particular, narrowing cast to subword types, since avx < 3 does not support truncation cast and only provides saturation cast instructions. We need to truncate the upper bits ourselves. For example, a cast from `int` to `byte` is done as follow
> 
> ```
> vpand dst, src, [external address mask]
> vpackusdw dst, dst
> vpermq dst, dst, 0x08
> vpackuswb dst, dst
> ```
> 
> For vector mask cast, we can get rid of the first masking and use the `vpackss`s instead, which removes the need to reference memory. Thanks.

I see, thanks! So would you like to provide the missing x86 backend implementation for `VectorMaskCast` ? If so we can use 
`VectorMaskCast` for all cases to simply the current codes? Thanks a lot!

**Comment by XiaohongGong** [comment]
> > Yes you are right, the code would be mostly the same, which means we can reuse the existing match rules to additionally match `VectorMaskCast` for those cases. For other cases, in particular, narrowing cast to subword types, since avx < 3 does not support truncation cast and only provides saturation cast instructions. We need to truncate the upper bits ourselves. For example, a cast from `int` to `byte` is done as follow
> > ```
> > vpand dst, src, [external address mask]
> > vpackusdw dst, dst
> > vpermq dst, dst, 0x08
> > vpackuswb dst, dst
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > For vector mask cast, we can get rid of the first masking and use the `vpackss`s instead, which removes the need to reference memory. Thanks.
> 
> I see, thanks! So would you like to provide the missing x86 backend implementation for `VectorMaskCast` ? If so we can use `VectorMaskCast` for all cases to simply the current codes? Thanks a lot!

Maybe I can refactor the codes in this patch, and add the same backend rules like `VectorCast`? And then you can create a followed-up patch to improve the x86 codegen if you like. WDYT?

**Comment by merykitty** [comment]
@XiaohongGong  I have created a PR against your branch, this only contains changes in the x86 backend to avoid any conflicts with the changes you may have made, could you have a look? Thanks

https://github.com/XiaohongGong/jdk/pull/2

**Comment by XiaohongGong** [comment]
> @XiaohongGong I have created a PR against your branch, this only contains changes in the x86 backend to avoid any conflicts with the changes you may have made, could you have a look? Thanks
> 
> [XiaohongGong#2](https://github.com/XiaohongGong/jdk/pull/2)

Sure, thanks a lot! I will take a look.

**Comment by XiaohongGong** [comment]
Hi @merykitty , thanks for the x86 backend implementation to the VectorMaskCast. I did a local refactor to the current code and tested with your backend patch, and no issue is found now. I guess there is no regression and I will run more benchmarks with it. BTW, I'd like to push the new changes to the github after https://github.com/openjdk/jdk/pull/9346 is merged because we have the dependence in the aarch64 backend rules.  Thanks!

**Comment by XiaohongGong** [comment]
Hi, the new commit refactors the vector mask cast implementation that we always use the VectorMaskCast op instead of VectorCast op for all platforms. The backend implementation for VectorMaskCast can be cheaper than VectorCast for some architectures like x86 avx. Please take a look at the changes. Thanks a lot!

**Comment by XiaohongGong** [comment]
> The performance improvement is impressive on x86 with avx2.
> 
> Can you clarify what tests did you do? Thanks.

Hi, thanks for looking at this PR! To test the x86 backend rules, I ran ` "test/jdk/jdk/incubator/vector,test/hotspot/jtreg/compiler/vectorapi"` group with `-XX:UseAVX=1, -XX:UseAVX=2, and -XX:UseAVX=3` separately on an AVX-512 machine. For the whole patch testing, I ran `hotspot::hotspot_all, jdk:tier1,tier2,tier3, langtools:tier1` on aarch64 and x86 machines. Thanks!

File: src/hotspot/cpu/x86/x86.ad [hunk 1]
```diff
@@ -8428,6 +8429,59 @@ instruct vmaskcast(vec dst) %{
...
+    BasicType dst_bt = Matcher::vector_element_basic_type(this);
+    __ vector_mask_cast($dst$$XMMRegister, $src$$XMMRegister, dst_bt, src_bt, vlen);
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct vmaskcast_avx1_32B_expand(vec dst, vec src, vec xtmp1, vec xtmp2) %{
+  predicate(Matcher::vector_length(n) == Matcher::vector_length(n->in(1)) &&
+            Matcher::vector_length_in_bytes(n) > Matcher::vector_length_in_bytes(n->in(1)) &&
+            UseAVX == 1 &&
```
Comment (DamonFool) [inline src/hotspot/cpu/x86/x86.ad:8452]: Since most x86 would be avx > 1, I would suggest testing `UseAVX == 1` first and then others.

File: src/hotspot/share/opto/vectornode.cpp [hunk 2]
```diff
@@ -1619,17 +1619,9 @@ Node* VectorUnboxNode::Ideal(PhaseGVN* phase, bool can_reshape) {
...
         bool is_vector_shuffle = vbox_klass->is_subclass_of(ciEnv::current()->vector_VectorShuffle_klass());
         if (is_vector_mask) {
+          // VectorUnbox (VectorBox vmask) ==> VectorMaskCast vmask
           const TypeVect* vmask_type = TypeVect::makemask(out_vt->element_basic_type(), out_vt->length());
-          if (in_vt->length_in_bytes() == out_vt->length_in_bytes() &&
-              Matcher::match_rule_supported_vector(Op_VectorMaskCast, out_vt->length(), out_vt->element_basic_type())) {
-            // Apply "VectorUnbox (VectorBox vmask) ==> VectorMaskCast (vmask)"
-            // directly. This could avoid the transformation ordering issue from
-            // "VectorStoreMask (VectorLoadMask vmask) => vmask".
-            return new VectorMaskCastNode(value, vmask_type);
```
Comment (DamonFool) [inline src/hotspot/share/opto/vectornode.cpp:1628]: Why do you change this code?
Is it a must to enable this optimization?

Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:1628]: For a vector mask unbox operation, if the input is a `VectorBox`, and the boxed value type is not equal to the unbox type, we need to do the type casting. Originally, we separate the type casting for two different cases:
1. if the vector element size in bytes is equal (e.g. casting between `int` and `float`), the conversion is:
```
  VectorUnbox (VectorBox vmask) ==> VectorMaskCast (vmask)
```
This doesn't need to emit any instructions.

2. otherwise (e.g. casting from `short` to `int`), the conversion is:
```
  VectorUnbox (VectorBox vmask) ==> VectorLoadMask (VectorStoreMask vmask)
```
This means we need to convert the short mask vector back to a boolean vector, and then casting the boolean vector to a int vector (i.e. `S->B->I`). 

With the new changes in this PR, `VectorMaskCast` supports the vector mask casting for all cases and all platforms (originally it only supports the first case for most platforms), we can implement the second cases simply with `VectorMaskCast`. That means we can optimize `S->B->I` to `S->I`. This saves the unnecessary narrowing and extending instructions due to the temporary boolean vector result.

Comment (DamonFool) [reply src/hotspot/share/opto/vectornode.cpp:1628]: > For a vector mask unbox operation, if the input is a `VectorBox`, and the boxed value type is not equal to the unbox type, we need to do the type casting. Originally, we separate the type casting for two different cases:
> 
> 1. if the vector element size in bytes is equal (e.g. casting between `int` and `float`), the conversion is:
> 
> ```
>   VectorUnbox (VectorBox vmask) ==> VectorMaskCast (vmask)
> ```
> 
> This doesn't need to emit any instructions.
> 
> 2. otherwise (e.g. casting from `short` to `int`), the conversion is:
> 
> ```
>   VectorUnbox (VectorBox vmask) ==> VectorLoadMask (VectorStoreMask vmask)
> ```
> 
> This means we need to convert the short mask vector back to a boolean vector, and then casting the boolean vector to a int vector (i.e. `S->B->I`).
> 
> With the new changes in this PR, `VectorMaskCast` supports the vector mask casting for all cases and all platforms (originally it only supports the first case for most platforms), we can implement the second cases simply with `VectorMaskCast`. That means we can optimize `S->B->I` to `S->I`. This saves the unnecessary narrowing and extending instructions due to the temporary boolean vector result.



So there are at least two optimizations in this pr:
1) Opt for FIRST_NONZERO
2) Opt for vector mask unbox operation

How about splitting them and doing the second one in a new PR？

Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:1628]: Thanks for the advice! Yes, this patch firstly fix the performance issue for FIRST_NONZERO and together with some clean up for the related codes. I considered to split the changes before. However, to think more, if the patch only fixes the performance issue, the final codes will look mess and we need to add the additional complex IR checks for the vector mask cast (please see the initial fixing changes here https://openjdk.github.io/cr/?repo=jdk&pr=9737&range=00). Unifying the usage to `VectorMaskCast` IR makes the final code clean. So I prefer to make the whole changes together.

Comment (DamonFool) [reply src/hotspot/share/opto/vectornode.cpp:1628]: > Thanks for the advice! Yes, this patch firstly fix the performance issue for FIRST_NONZERO and together with some clean up for the related codes. I considered to split the changes before. However, to think more, if the patch only fixes the performance issue, the final codes will look mess and we need to add the additional complex IR checks for the vector mask cast (please see the initial fixing changes here https://openjdk.github.io/cr/?repo=jdk&pr=9737&range=00). Unifying the usage to `VectorMaskCast` IR makes the final code clean. So I prefer to make the whole changes together.


I'm not sure if there would be some side effect about the vector unbox code change.

Since the original goal of this pr is optimizing `FIRST_NONZERO`, it might be better to do the others (code cleanup, unifying the usage of VectorMaskCast) in a separate pr to lower the risk.
Thanks.

Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:1628]: OK, I will split the patch based on your suggestion. Thanks again for the review!

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.hpp [hunk 3]
```diff
@@ -332,6 +332,11 @@
   void vector_unsigned_cast(XMMRegister dst, XMMRegister src, int vlen_enc,
                             BasicType from_elem_bt, BasicType to_elem_bt);
 
+  void vector_mask_cast(XMMRegister dst, XMMRegister src, BasicType dst_bt, BasicType src_bt, int vlen);
+
+  void vector_mask_cast_with_tmp(XMMRegister dst, XMMRegister src, XMMRegister xtmp1,
```
[the remark this answers is not in the document]
Comment (XiaohongGong) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.hpp:337]: Will fix this in a separate patch. Thanks!

File: src/hotspot/cpu/x86/x86.ad (hunk 1 above)
Comment (XiaohongGong) [reply src/hotspot/cpu/x86/x86.ad:8452]: Will fix this in a separate patch, thanks!

Comment (merykitty) [reply src/hotspot/cpu/x86/x86.ad:8452]: I think the general cases should come before the particular ones. These 2 rules are for a corner case where float or double masks with 256-bit width are involved. Thanks.

Comment (DamonFool) [reply src/hotspot/cpu/x86/x86.ad:8452]: > I think the general cases should come before the particular ones. These 2 rules are for a corner case where float or double masks with 256-bit width are involved. Thanks.

For modern x86 platforms, UseAVX would be >= 2, these 2 rules can be filtered out as soon as possible if we test  `UseAVX == 1` at the beginning.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (merykitty) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.hpp:337]: I think it's fine, I see we don't name methods in `C2_MacroAssembler` with the same name, except when it is the addessliteral version that would likely to emit the same thing.

**Comment by XiaohongGong** [comment]
Hi, the new commit reverted the changes to unify the vector mask cast operation with `VectorMaskCast` op, so that the whole patch can focus on fixing the performance issue on `DoubleMaxVector.FIRST_NONZERO`. A followed-up PR will specially refactor the vector mask casting. Please kindly review and share feedback. Thanks!

**Comment by DamonFool** [comment]
Thanks for the update.

May I ask can we do it like this?
```
diff --git a/src/hotspot/share/opto/vectorIntrinsics.cpp b/src/hotspot/share/opto/vectorIntrinsics.cpp
index 66bacf0..c4e807a 100644
--- a/src/hotspot/share/opto/vectorIntrinsics.cpp
+++ b/src/hotspot/share/opto/vectorIntrinsics.cpp
@@ -2494,8 +2494,9 @@ bool LibraryCallKit::inline_vector_convert() {
       new_elem_bt_from = elem_bt_from == T_FLOAT ? T_INT : T_LONG;
     }
     int cast_vopc = VectorCastNode::opcode(new_elem_bt_from, !is_ucast);
+    bool no_vec_cast_check = is_mask && (type2aelembytes(elem_bt_from) == type2aelembytes(elem_bt_to));
     // Make sure that cast is implemented to particular type/size combination.
-    if (!arch_supports_vector(cast_vopc, num_elem_to, elem_bt_to, VecMaskNotUsed)) {
+    if (!no_vec_cast_check && !arch_supports_vector(cast_vopc, num_elem_to, elem_bt_to, VecMaskNotUsed)) {
       if (C->print_intrinsics()) {
         tty->print_cr("  ** not supported: arity=1 op=cast#%d/3 vlen2=%d etype2=%s ismask=%d",
                       cast_vopc,
```

**Comment by XiaohongGong** [comment]
> Thanks for the update.
> 
> May I ask can we do it like this?
> 
> ```
> diff --git a/src/hotspot/share/opto/vectorIntrinsics.cpp b/src/hotspot/share/opto/vectorIntrinsics.cpp
> index 66bacf0..c4e807a 100644
> --- a/src/hotspot/share/opto/vectorIntrinsics.cpp
> +++ b/src/hotspot/share/opto/vectorIntrinsics.cpp
> @@ -2494,8 +2494,9 @@ bool LibraryCallKit::inline_vector_convert() {
>        new_elem_bt_from = elem_bt_from == T_FLOAT ? T_INT : T_LONG;
>      }
>      int cast_vopc = VectorCastNode::opcode(new_elem_bt_from, !is_ucast);
> +    bool no_vec_cast_check = is_mask && (type2aelembytes(elem_bt_from) == type2aelembytes(elem_bt_to));
>      // Make sure that cast is implemented to particular type/size combination.
> -    if (!arch_supports_vector(cast_vopc, num_elem_to, elem_bt_to, VecMaskNotUsed)) {
> +    if (!no_vec_cast_check && !arch_supports_vector(cast_vopc, num_elem_to, elem_bt_to, VecMaskNotUsed)) {
>        if (C->print_intrinsics()) {
>          tty->print_cr("  ** not supported: arity=1 op=cast#%d/3 vlen2=%d etype2=%s ismask=%d",
>                        cast_vopc,
> ```

Thanks for the advice! The new commit has fixed this issue. Could you please check whether it is ok for you? Thanks!

File: src/hotspot/share/opto/vectorIntrinsics.cpp [hunk 4]
```diff
@@ -2493,13 +2494,18 @@ bool LibraryCallKit::inline_vector_convert() {
     if (is_mask && is_floating_point_type(elem_bt_from)) {
       new_elem_bt_from = elem_bt_from == T_FLOAT ? T_INT : T_LONG;
     }
+    if (is_mask && is_floating_point_type(elem_bt_to)) {
+      new_elem_bt_to = elem_bt_to == T_FLOAT ? T_INT : T_LONG;
```
Comment (DamonFool) [inline src/hotspot/share/opto/vectorIntrinsics.cpp:2498]: Why we need this change?

File: src/hotspot/share/opto/vectorIntrinsics.cpp [hunk 5]
```diff
@@ -2493,13 +2494,18 @@ bool LibraryCallKit::inline_vector_convert() {
...
+    }
     int cast_vopc = VectorCastNode::opcode(new_elem_bt_from, !is_ucast);
-    // Make sure that cast is implemented to particular type/size combination.
-    if (!arch_supports_vector(cast_vopc, num_elem_to, elem_bt_to, VecMaskNotUsed)) {
+
+    // Make sure that vector cast is implemented to particular type/size combination.
+    bool no_vec_cast_check = is_mask && (src_type->isa_vectmask() ||
```
Comment (DamonFool) [inline src/hotspot/share/opto/vectorIntrinsics.cpp:2503]: To make it safer, let's also check `dst_type->isa_vectmask()` like the following.
```
      if (is_mask) {
        if ((dst_type->isa_vectmask() && src_type->isa_vectmask()) ||
            (type2aelembytes(elem_bt_from) == type2aelembytes(elem_bt_to))) {
          op = gvn().transform(new VectorMaskCastNode(op, dst_type));
        } else {
          op = VectorMaskCastNode::makeCastNode(&gvn(), op, dst_type);
        }
      }
```

File: src/hotspot/share/opto/vectorIntrinsics.cpp (hunk 4 above)
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectorIntrinsics.cpp:2498]: Before `VectorCast` is used for vector mask casting, the src floating point type has been casted to an integral type that has the same element size. Please see the codes here: https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/vectornode.cpp#L1741. We'd better to check the right type of the generated vector cast op.

File: src/hotspot/share/opto/vectorIntrinsics.cpp (hunk 5 above)
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectorIntrinsics.cpp:2503]: From the check https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/vectorIntrinsics.cpp#L2485, we know the basic mask type of dst and src must be the same. So just checking `src_type` is ok here. Besides, this part of changes will be removed finally after I unify the vector mask cast operation.

Comment (DamonFool) [reply src/hotspot/share/opto/vectorIntrinsics.cpp:2503]: But the mismatch seems strange to me.

Maybe, you can 
1) Add `dst_type->isa_vectmask()` check here, or
2) Also delete `dst_type->isa_vectmask()`  (https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/vectorIntrinsics.cpp#L2551)

File: src/hotspot/share/opto/vectorIntrinsics.cpp (hunk 4 above)
Comment (DamonFool) [reply src/hotspot/share/opto/vectorIntrinsics.cpp:2498]: OK, seems fine to me.
Thanks.

File: src/hotspot/share/opto/vectorIntrinsics.cpp (hunk 5 above)
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectorIntrinsics.cpp]: I'v added the dst_type check in the latest commit. Thanks!

**Review by DamonFool — APPROVED** [review]
LGTM
Thanks for the update.

**Comment by XiaohongGong** [comment]
> LGTM Thanks for the update.

Thanks so much for the review!

**Comment by XiaohongGong** [comment]
Hi @jatin-bhateja, could you please help to take a look at this PR? Thanks so much for your time!

**Review by theRealELiu — APPROVED** [review]
LGTM.

**Comment by XiaohongGong** [comment]
ping again. Could anyone else please help to take a look at this PR? Thanks a lot!

**Comment by XiaohongGong** [comment]
The GHA failure is not related to this PR. So may I get an approve from someone else? Thanks in advance!

File: src/hotspot/share/opto/vectorIntrinsics.cpp [hunk 6]
```diff
@@ -2493,13 +2494,19 @@ bool LibraryCallKit::inline_vector_convert() {
...
+      new_elem_bt_to = elem_bt_to == T_FLOAT ? T_INT : T_LONG;
+    }
     int cast_vopc = VectorCastNode::opcode(new_elem_bt_from, !is_ucast);
-    // Make sure that cast is implemented to particular type/size combination.
-    if (!arch_supports_vector(cast_vopc, num_elem_to, elem_bt_to, VecMaskNotUsed)) {
+
+    // Make sure that vector cast is implemented to particular type/size combination.
+    bool no_vec_cast_check = is_mask &&
+                             ((src_type->isa_vectmask() && dst_type->isa_vectmask()) ||
+                              type2aelembytes(elem_bt_from) == type2aelembytes(elem_bt_to));
```
Comment (jatin-bhateja) [inline src/hotspot/share/opto/vectorIntrinsics.cpp:2505]: Same check is existing at https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/vectorIntrinsics.cpp#L2551, we can do some factorization.

File: src/hotspot/share/opto/vectorIntrinsics.cpp [hunk 7]
```diff
@@ -2541,12 +2548,10 @@ bool LibraryCallKit::inline_vector_convert() {
...
 
-      op = gvn().transform(new VectorReinterpretNode(op,
-                                                     src_type,
-                                                     TypeVect::make(elem_bt_from,
-                                                                    num_elem_for_resize)));
+      const TypeVect* resize_type = TypeVect::make(elem_bt_from, num_elem_for_resize);
+      op = gvn().transform(new VectorReinterpretNode(op, src_type, resize_type));
       op = gvn().transform(VectorCastNode::make(cast_vopc, op, elem_bt_to, num_elem_to));
-    } else {
+    } else { // num_elem_from == num_elem_to
```
Comment (jatin-bhateja) [inline src/hotspot/share/opto/vectorIntrinsics.cpp:2554]: There is one comment unrelated to this patch but since your patch touches this function may be we can address it.
Call to VectorMaskCastNode::makeCastNode in else block at following line
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/vectorIntrinsics.cpp#L2555
should be able to handle the true block. 
But, newly created IR node on following location https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/vectornode.cpp#L1761 should be passed through gvn transform before returning.

**Review by jatin-bhateja — COMMENTED** [review]
Thanks @XiaohongGong  for looking into this,  patch is showing significant speedups for AVX2 and AVX512 KNL targets.

File: src/hotspot/share/opto/vectorIntrinsics.cpp (hunk 7 above)
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectorIntrinsics.cpp:2554]: Yes, you are right. That's also what is the first version of this PR (please see: https://openjdk.github.io/cr/?repo=jdk&pr=9737&range=00).  And since @merykitty  suggested to unify the whole vector mask cast operations with `VectorMaskCastNode`, I prepare a follow-up patch to specifically do this refactorization (please see: https://github.com/openjdk/jdk/pull/10192). And as suggested by @DamonFool , we only fix the `FIRST_NONZERO` performance issue in this PR.

File: src/hotspot/share/opto/vectorIntrinsics.cpp (hunk 6 above)
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectorIntrinsics.cpp:2505]: Yes, we can. But consider these codes will be optimized out in the follow-up unify patch (see https://github.com/openjdk/jdk/pull/10192), do you think it's necessary to do some simple factorization here? If it's necessary, I can change the codes. Thanks!

File: src/hotspot/share/opto/vectorIntrinsics.cpp [hunk 8]
```diff
@@ -2482,7 +2482,8 @@ bool LibraryCallKit::inline_vector_convert() {
   // through a vector or predicate register.
   if (is_mask &&
       ((src_type->isa_vectmask() == NULL && dst_type->isa_vectmask()) ||
-       (dst_type->isa_vectmask() == NULL && src_type->isa_vectmask()))) {
+       (dst_type->isa_vectmask() == NULL && src_type->isa_vectmask()) ||
+       num_elem_from != num_elem_to)) {
```
[the remark this answers is not in the document]
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectorIntrinsics.cpp:2486]: Yes, it is. Just want to double check in the compiler side, or it maybe confusing if someone is not familiar with the vector api java codes.

File: src/hotspot/share/opto/vectorIntrinsics.cpp (hunk 7 above)
Comment (jatin-bhateja) [reply src/hotspot/share/opto/vectorIntrinsics.cpp:2554]: Yes, I just noticed that #10192 is addressing above two comments

Comment (XiaohongGong) [reply src/hotspot/share/opto/vectorIntrinsics.cpp:2554]: Yeah, thanks so much for looking at that PR as well!

File: src/hotspot/share/opto/vectorIntrinsics.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (jatin-bhateja) [reply src/hotspot/share/opto/vectorIntrinsics.cpp:2486]: I see, may be we can just add a comment instead of injecting code.

[the remark this answers is not in the document]
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectorIntrinsics.cpp:2486]: Yeah, agree to you. Or maybe as assert under `if(num_elem_from < num_elem_to)` and `if(num_elem_from > num_elem_to)` ?  I'm sorry that I didn't find a better place to add the comment. Any suggestions? For me, it's better to add the comment inside the `if(is_cast)` block since it has different paths based on the `num_elem_from` and `num_elem_to`.

[the remark this answers is not in the document]
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectorIntrinsics.cpp:2486]: Something like: `assert(!is_mask, "mask cast needs the same elem num ")`

[the remark this answers is not in the document]
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectorIntrinsics.cpp:2486]: Hi, I'd like the add the assertion in the unify patch instead of this. So I will revert this change here. Thanks!

[the remark this answers is not in the document]
Comment (jatin-bhateja) [reply src/hotspot/share/opto/vectorIntrinsics.cpp:2486]: Yes, assert (!is_mask || (is_mask && num_elem_from == num_elem_to),  "").

[the remark this answers is not in the document]
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectorIntrinsics.cpp:2486]: Looks good. Thanks!

[the remark this answers is not in the document]
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectorIntrinsics.cpp]: The new commit reverted this changes. I will add the assertion in https://github.com/openjdk/jdk/pull/10192.

**Review by DamonFool — APPROVED** [review]
Still looks good to me.
Thanks.

**Review by jatin-bhateja — APPROVED** [review]
